File: OOPinC/stackDataHiding/Stack.c
#include <stdlib.h>
#include <stdio.h>
#include "Stack.h"

/* Hidden within this file. */
static int top_ , size_ ;
static T *stack_ ;

int create (int size) {
	top_= 0; size_ = size;
	stack_ = malloc (size * sizeof (T));
	return stack_ == 0 ? -1 : 0;
}
void destroy (void) { free ((void *) stack_ ); }
void push (T item) { stack_ [top_ ++] = item; }
void pop (T *item) { *item = stack_ [--top_ ]; }
void top (T *item) { *item = stack_ [top_ -1]; }
int is_empty (void) { return top_ == 0; }
int is_full (void) { return top_ == size_ ; }
void print(){ for(int i=0; i < size_; i++) printf("%d ",stack_[i]); }

File: stack2Java/Stack.java
package stack2Java;
import java.lang.reflect.Array;  //generic array object "meta-type"

class Stack<T> {	
    private int top, size;
	private T[] stack;
    private String type;

   public Stack(Class<T> type, int size){
        this.top = 0; 
        this.size = size; 
        this.stack = (T[]) Array.newInstance(type, size); // class type == T  should match
    }
	
    public void push (T item) throws StackOverflowException{
        if (this.is_full()) throw new StackOverflowException(this); 
        this.stack[this.top++] = item;
    }
	public T pop ()throws StackUnderflowException{
        if (this.is_empty()) throw new StackUnderflowException(this); 
        return 	this.stack[--this.top];

    }
	public boolean is_empty (){ return this.top == 0;}
	public boolean is_full (){ return this.top == this.size;}
		//…
       
    void setType(String type){this.type = type;}
    String getType(){return this.type;}
	
};
